### ISSUE TITLE
Still missing some of the thermal relief paths

### ISSUE THREAD

**Issue by a7hybnj2** [pcb2gcode/pcb2gcode#196]
@eyal0 
 I decided since this is technically a new issue.

[Link to output](https://github.com/a7hybnj2/KicadStuffs/tree/master/pcb2gcode_experience/With%20boost%201.68%20and%20pcb2gcode%209ff6a3b%20)

The output is so close to being perfect. I don't understand why it would be missing the thermal reliefs that it is.

**Comment by eyal0** [comment]
Discussion moved from here: https://github.com/pcb2gcode/pcb2gcode/issues/195#issuecomment-412120493

**Comment by eyal0** [comment]
Okay, off the top of my head: What is your setting for offset? Maybe the offset is bigger than the size of the thermal? If the milling bit is bigger than the thermal relief, the thermal relief will go away. To test this, try setting your your offset to 0 temporarily.

As in, --offset=0

**Comment by a7hybnj2** [comment]
Same results with --offset 0

[offset 0](https://github.com/a7hybnj2/KicadStuffs/tree/master/pcb2gcode_experience/With%20boost%201.68%20and%20pcb2gcode%209ff6a3b%20)

**Comment by eyal0** [comment]
I'll try it out on my computer and see if I can figure it out.

**Comment by eyal0** [comment]
Okay, I'm seeing the problem, too.  :disappointed: I looked at:

1. Running gerbv on the file: looks good
2. outp0_original_back.svg: missing thermals
3. back.ngc: missing thermals

So it seems that the step that converts from gerber to boost geometry is broken.  But the step from boost geometry to gcode is fine.

That's a bummer because I have very little experience in the former.  But the good news is that I get the opportunity to learn it.  Yay!  So, I'll get started on that.  Might take a little while.

In your outp0_original_front.png file all the thermal reliefs are there.  But in my outp_original file, they are not there.  Also, mine is .svg and yours is png.  Can you tell me how you got your .png file?

**Comment by a7hybnj2** [comment]
Yea, I did not do a good job of organizing the output and deleting/renaming old outputs. This directory is a fresh output. I started just with the .grb file and the millproject file then I just ran `pcb2gode` while in that directory and the remaining files are the result.

[Everything Current](https://github.com/a7hybnj2/KicadStuffs/tree/master/pcb2gcode_experience/Fresh%20Test)

**Comment by eyal0** [comment]
Okay, that's similar to my experience.  I'll look into it.

**Comment by eyal0** [comment]
I think that I have a fix for your problem.  I guessed at the issue: bad cutins algorithm, and I was right.  I think.  Would you test it out and let me know how you like the results?

Either download this: https://github.com/eyal0/pcb2gcode/archive/better_cutins.zip

Or checkout the branch eyal0/better_cutins

Let me know if you have any trouble doing either of those.

Then just compile it like normal.  Let me know how it runs or if it seems extra slow or anything.  In my testing, I see many more thermal reliefs.  Let me know if you see any more of them missing!

**Comment by a7hybnj2** [comment]
Totally works!  
Issue: it did not flip the gcode for output. See screenshop.
Question: is there a way to add file-ext=*.nc to millproject instead of the output being *ngc?

https://github.com/a7hybnj2/KicadStuffs/tree/master/180815_better_cutins

**Comment by eyal0** [comment]
What do you mean not flipped?

In the source gerber, using gerbv, I see the squrae thermal reliefs on the left.  In the .ngc file, I see them on the right.  So it seems okay to me!  What did you see and what do you expect?

I thought about a file extension feature but it's seems too specialized.  It's too easy to write a one-line script in UNIX or use some other tool to convert all *.ngc to *.nc .

Did you try the --back-output option?

**Comment by a7hybnj2** [comment]
Your right! I was thinking the the generated .svg files out be in the orientation I wanted.

No, I did not try the "--back-output" because I never saw that in the manual on the wiki.

And I guess all of the x-vals are (-) because that is how you flip it? So, if I were doing a double sided pcb I would have to figure out something other then my alignment pins. Or maybe flip the gerber then mark it as --front...

**Comment by eyal0** [comment]
Better to just set the mirror-axis to whatever line you want to use for flipping. Try mirror-axis=20mm and see the results.

**Comment by a7hybnj2** [comment]
I will give that a try. I am doing some offset + extra passes tests now that based on my first tests do not seem to be working. But maybe I am misunderstanding how it is to work.

If I set offset=0 I get a line. If I set offset=0.5 extra-passes=5 will any of those lines be the same as offset 0 pass? How would you make the isolation larger with multiple passes?

**Comment by a7hybnj2** [comment]
Hey @eyal0 can you double check the manual and man page? I did not find the mirror-axis in there either. Maybe I am missing out on some primo flags!

**Comment by eyal0** [comment]
The documentation is old. I should delete it. Only the wiki is completely accurate and it is missing lots.

Mirror-absolute is deprecated. It never worked correctly so I got rid of it.

The best way to know all the options is with --help

Offest is how far from the trace to mill. So if your tool is 1mm wide then 0.5 mm offset is a good choice because the milling will happen 0.5mm away from the trace and will give you a nice outline. If extra-passes is more than 0, there will be more milling surrounding that first pass and each will be an additional 0.5 away from the previous. So you can have a milling at 0.5mm away and then 1.0mm, then 1.5,2.0,etc

**Comment by eyal0** [comment]
Mirror-absolute must always be true now.

**Comment by a7hybnj2** [comment]
I am making headway. Your explanation of offset makes sense.

The last thing (for now) I am struggling with is an 'area clear' on the drill file. I am trying to get mill-drill with the correct parameters but I cannot get it to cut. Will you take a look at this?

https://github.com/a7hybnj2/KicadStuffs/tree/master/180815_better_cutins

**Comment by eyal0** [comment]
What do you mean by area clear? Like, you want an area of the board to have all it's copper removed? We don't have a feature like that now because I'm not sure how to implement it. If you can think of a way to for the user to specify which area should be completely cleared of copper, let me know!

Milldrill is for using a milling bit to drill the holes. Usually holes are drilled by a drill, straight through. Milldrill uses a milling bit and makes a few passes.

The wiki covers it.

**Comment by eyal0** [comment]
https://github.com/pcb2gcode/pcb2gcode/wiki/Options:-Drilling

**Comment by a7hybnj2** [comment]
Yea, I wasn’t very clear... area clear is essentially what the milldrill setting does, area clear inside the circles. And I am trying to do it on the drill file... which is just a bunch of tiny circles. I just wasn’t getting cuts from it. It would move to each location but nothing got cut.

**Comment by eyal0** [comment]
Oh, now I understand your interesting solution! Yeah, that's one way to do it!

A drill file is usually a bunch of holes that you want to drill clear through to the other side of the board. Both regular drilling and milldrilling can do that. But milldrilling does it by milling a bunch of increasingly deeper rings. The bigger the ring, the bigger the hole. Makes sense?

Now, pcb2gcode is an isolation PCB creation method. That means that we don't remove all the unneeded copper, we just mill away enough copper to isolate the traces from other traces. If you used a chemical bath instead then you might get rid of all the copper that you don't need.

So since it's only doing isolation, parts of the PCB that don't have copper actually still have copper when the milling is done. For most people, this is fine. If you've made a big ground plane it might not even make a difference.

But if you've got an antenna or something that would be blocked by the copper then you need more cleared. One option is to set extra-passes to a really high number. It will make concentric rings around your work and clear out a lot of copper but it'll take a long time.

If I could think of a good way for users to specify where to clear copper and where to just isolate, then I could add the feature. But that's not a normal part of PCB design so I don't know how it should be done.

**Comment by eyal0** [comment]
I still don't totally understand the milldrill issues.  Are you using a negative offset because you want the milldrill to be used for clearing copper and not for actual milldrilling?

**Comment by a7hybnj2** [comment]
@eyal0 No. I am trying to milldrill the drill file on my pcb. So, I am trying to use it correctly. It generates a ngc file and moves to each hole location but doesn't cut.

Since, I could not figure it out I started 'misusing' "--back mydrillfile.drl", "--offset -.075", "--extra-passes 10" in order to clear out the drill holes.

When I said 'area clear' previously I was refering to the area inside each tiny drill circle. I have no desire to remove extra copper and I like my huge ground planes.

**Comment by eyal0** [comment]
Ah, okay. Then I'll take a look at your settings and see what is going on

**Comment by a7hybnj2** [comment]
I will give it a shot
